File: examples/c/perplexity.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "fastllama.h"

int main() {
    struct llama_model_context_args args = llama_create_default_context_args();
    args.n_threads = 16;
    args.n_batch = 512; // number of samples per batch
    args.n_ctx = 512;
    args.n_keep = 200;

    struct llama_model_context* model_ctx = llama_create_context(args);
    
    if (!llama_load_model(model_ctx, LLAMA_7B, "./models/7B/ggml-model-q4_0.bin")) {
        return 1;
    }

    // Use some text file for testing that has at least 8000 characters
    FILE* file = fopen("./test.txt", "r");
    if (file == NULL) {
        printf("Failed to open file");
        return 1;
    }

    // Get 8000 characters for testing
    char buffer[8000] = {0};
    int read_count = fread(&buffer, sizeof(char), sizeof(buffer) - 1, file);
    if (read_count == 0) {
        printf("Failed to read file");
        return 1;
    }

    float final_result = llama_perplexity(model_ctx, buffer);
    printf("Total Perplexity: %.4f", final_result);
    
    return 0;
}

File: examples/python/perplexity.py
from build.fastllama import Model, ModelKind

MODEL_PATH = "./models/7B/ggml-model-q4_0.bin"

model = Model(
        id=ModelKind.ALPACA_LORA_7B,
        path=MODEL_PATH, #path to model
        num_threads=16, #number of threads to use
        n_ctx=512, #context size of model
        last_n_size=16, #size of last n tokens (used for repetition penalty) (Optional)
        n_batch=512, #number of batches to use
    )

# Perplexity caculation on a file named test.txt (8000 bytes)
with open("test.txt", "r") as f:
    data = f.read(8000)
    total_perplexity = model.perplexity(data)
    print(f"Total Perplexity: {total_perplexity:.4f}")
